File: gaussian_elimination.py
def gaussian_elimination(M: list, S: list):
    """Returns the solution of a system with 3 unknowns. It utilizes the Gaussian elimination method.

    Args:
        - {list} M : Matrix with coefficients
        - {list} S : Matrix with numbers
    """

    # Rows of the matrix
    R_1 = M[0]
    R_2 = M[1]
    R_3 = M[2]

    # First step
    if R_1[0] == 0:
        # swap rows if the pivot is 0
        R_1, R_2 = R_2, R_1
        S[0], S[1] = S[1], S[0]
    else:
        p = R_1[0] # pivot
        S[1] = S[1]-R_2[0]/p*S[0]
        S[2] = S[2]-R_3[0]/p*S[0]
        R_2 = [R_2[i] - (R_2[0] / p) * R_1[i] for i in range(3)]
        R_3 = [R_3[i] - (R_3[0] / p) * R_1[i] for i in range(3)]

    # second step
    if R_2[1] == 0:
        # swaps the lines if the pivot is 0
        R_2, R_3 = R_3, R_2
        S[1], S[2] = S[2], S[1]
    else:
        p = R_2[1] # pivot
        # change the solution matrix
        S[2] = S[2]-R_3[1]/p*S[1]
        R_3 = [R_3[i]-(R_3[1]/p)*R_2[i] for i in range(3)]

    # solution step
    z = S[2]/R_3[2]
    y = (S[1]-R_2[2]*z)*1/R_2[1]
    x = (S[0]-R_1[1]*y-R_1[2]*z)*1/R_1[0]

    # --- solutions --- #
    solution = (
        "\033[35mThe solution of the system is:\n" +
        f"\033[31mx = {x}\n" +
        f"\033[34my = {y}\n" +
        f"\033[33mz = {z}"
    )
    return solution

M = [[1, 2, -1], [2, 3, 1], [1, 1, 1]]
S = [4, 17, 9]
print(gaussian_elimination(M, S))

File: gaussian_elimination.rb
require 'colorize'


def gaussian_elimination(m, s)
    # Returns the solution of a system with 3 unknowns. It utilizes the Gaussian elimination method.
    #
    # Arguments:
    #   - {Array} m : matrix with coefficients
    #   - {Array} s : matrix with numbers
    #

    # Rows of the matrix
    r_1 = m[0]
    r_2 = m[1]
    r_3 = m[2]

    # First step
    if r_1[0] == 0
        # swap rows if the pivot is 0
        r_1, r_2 = r_2, r_1
        s[0], s[1] = s[1], s[0]
    else
        p = r_1[0] # pivot
        s[1] = s[1]-r_2[0]/p*s[0]
        s[2] = s[2]-r_3[0]/p*s[0]
        
        # with_index (Ruby) <=> enumerate (Python)
        # value = r_2[i] (0...4).each do |i|
        r_2 = r_2.map.with_index {|value, i| value - (r_2[0] / p) * r_1[i]}
        r_3 = r_3.map.with_index {|value, i| value - (r_3[0] / p) * r_1[i]}
    end

    # second step
    if r_2[1] == 0
        # swaps the lines if the pivot is 0
        r_2, r_3 = r_3, r_2
        s[1], s[2] = s[2], s[1]
    else
        p = r_2[1] # pivot
        # change the solution matrix
        s[2] = s[2]-r_3[1]/p*s[1]
        r_3 = r_3.map.with_index {|value, i| value - (r_3[1] / p) * r_2[i]}
    end

    # solution step
    z = s[2]/r_3[2]
    y = (s[1]-r_2[2]*z)*1/r_2[1]
    x = (s[0]-r_1[1]*y-r_1[2]*z)*1/r_1[0]

    # --- solutions --- #
    solution = (
        "The solution of the system is:\n".colorize(:magenta) +
        "x = #{x}\n".colorize(:red) +
        "y = #{y}\n".colorize(:blue) +
        "z = #{z}".colorize(:yellow)
    )
    return solution
end 

m = [[1, 2, -1], [2, 3, 1], [1, 1, 1]]
s = [4, 17, 9]
puts gaussian_elimination(m, s)